### PR TITLE
planner: add two functions to get and set cached plan from the plan cache

### PR DESCRIPTION
**Pull request by qw4990**
### What problem does this PR solve?

Issue Number: ref #36598

Problem Summary: planner: add two functions to get and set cached plan from the plan cache

### What is changed and how it works?

No logical change, just refactor, add two functions to get and set cached plan from the plan cache

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: planner/core/plan_cache_utils.go [hunk 1]
```diff
@@ -39,6 +39,42 @@ var (
...
+	candidates := val.([]*PlanCacheValue)
+	for i, candidate := range candidates {
+		if candidate.varTypesUnchanged(plan.ParamTypes) {
+			// hit an existing cached plan
+			candidates[i] = plan
+			return
+		}
+	}
+	// add to current candidate list
+	// TODO: limit the candidate list length
```
Comment (qw4990) [inline planner/core/plan_cache_utils.go:73]: I'll limit it in the next PR.

File: planner/core/plan_cache.go [hunk 2]
```diff
@@ -208,55 +207,42 @@ func getPointQueryPlan(stmt *ast.Prepared, sessVars *variable.SessionVars, stmtC
...
+func getGeneralPlan(sctx sessionctx.Context, cacheKey kvcache.Key,
+	bindSQL string, stmt *PlanCacheStmt, paramTypes []*types.FieldType) (Plan,
 	[]*types.FieldName, bool, error) {
 	sessVars := sctx.GetSessionVars()
 	stmtCtx := sessVars.StmtCtx
 
-	if cacheValue, exists := sctx.PreparedPlanCache().Get(cacheKey); exists {
-		if err := CheckPreparedPriv(sctx, stmt, is); err != nil {
-			return nil, nil, false, err
-		}
```
Comment (Reminiscent) [inline planner/core/plan_cache.go:220]: It seems we didn't check this part in the new codebase.

**Review by Reminiscent — APPROVED** [review]
Rest LGTM.

File: planner/core/plan_cache.go (hunk 2 above)
Comment (qw4990) [reply planner/core/plan_cache.go:220]: Nice catch.

Comment (qw4990) [reply planner/core/plan_cache.go:220]: Updated, PTAL.

**Review by xuyifangreeneyes — APPROVED** [review]

**Pull request merged**